File: WrapperCielo24/WrapperCielo24/Actions.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.IO;
using System.Security.Principal;
using System.Net;

namespace WrapperCielo24
{
    public class Actions
    {
        private const string BASE_URI = "https://sandbox-dev.cielo24.com";
        public const int VERSION = 1;

        private const string LOGIN_PATH = "/api/account/login";
        private const string LOGOUT_PATH = "/api/account/logout";
        private const string UPDATE_PASSWORD_PATH = "/api/account/update_password";
        private const string GENERATE_API_KEY_PATH = "/api/account/generate_api_key";
        private const string REMOVE_API_KEY_PATH = "/api/account/remove_api_key";
        private const string CREATE_JOB_PATH = "/api/job/new";
        private const string AUTHORIZE_JOB_PATH = "/api/job/authorize";
        private const string DELETE_JOB_PATH = "/api/job/del";
        private const string GET_JOB_INFO_PATH = "/api/job/list";
        private const string GET_JOB_LIST_PATH = "/api/job/list";
        private const string ADD_MEDIA_TO_JOB_PATH = "/api/job/add_media";
        private const string ADD_EMBEDDED_MEDIA_TO_JOB_PATH = "/api/job/add_media_url";
        private const string GET_MEDIA_PATH = "/api/job/media";
        private const string GET_TRANSCRIPTION_PATH = "/api/job/get_transcript";
        private const string GET_CAPTION_PATH = "/api/job/get_caption";
        private const string GET_ELEMENT_LIST_PATH = "/api/job/get_elementlist";
        private const string GET_LIST_OF_ELEMENT_LISTS_PATH = "/api/job/list_elementlists";

        /// ACCESS CONTROL ///
        
        /* Performs a Login action. If useHeaders is true, puts username and password into HTTP headers */
        public Guid Login(string username, string password, bool useHeaders=false)
        {
            this.AssertArgument(username, "Username");
            this.AssertArgument(password, "Password");

            Dictionary<string, string> queryDictionary = new Dictionary<string, string>();
            Dictionary<string, string> headers = new Dictionary<string, string>();
            queryDictionary.Add("v", VERSION.ToString());

            if (!useHeaders)
            {
                queryDictionary.Add("username", username);
                queryDictionary.Add("password", password);
            }
            else
            {
                headers.Add("x-auth-user", username);
                headers.Add("x-auth-key", password);
            }

            Uri requestUri = Utils.BuildUri(BASE_URI, LOGIN_PATH, queryDictionary);
            WebUtils web = new WebUtils();

            string serverResponse = web.HttpRequest(requestUri, HttpMethod.GET, WebUtils.BASIC_TIMEOUT, headers);
            Dictionary<string, string> response = Utils.DeserializeDictionary(serverResponse);

            return new Guid(response["ApiToken"]);
        }

        /* Performs a Login action. If useHeaders is true, puts securekey into HTTP headers */
        public Guid Login(string username, Guid securekey, bool useHeaders = false)
        {
            this.AssertArgument(username, "Username");

            Dictionary<string, string> queryDictionary = new Dictionary<string, string>();
            Dictionary<string, string> headers = new Dictionary<string, string>();
            queryDictionary.Add("v", VERSION.ToString());

            if (!useHeaders)
            {
                queryDictionary.Add("username", username);
                queryDictionary.Add("securekey", securekey.ToString("N"));
            }

            else
            {
                headers.Add("x-auth-user", username);
                headers.Add("x-auth-securekey", securekey.ToString("N"));
            }

            Uri requestUri = Utils.BuildUri(BASE_URI, LOGIN_PATH, queryDictionary);
            WebUtils web = new WebUtils();

            string serverResponse = web.HttpRequest(requestUri, HttpMethod.GET, WebUtils.BASIC_TIMEOUT, headers);
            Dictionary<string, string> response = Utils.DeserializeDictionary(serverResponse);
            
            return new Guid(response["ApiToken"]);
        }

        /* Performs a Logout action */
        public void Logout(Guid apiToken)
        {
            Dictionary<string, string> queryDictionary = this.InitAccessReqDict(apiToken);
            Uri requestUri = Utils.BuildUri(BASE_URI, LOGOUT_PATH, queryDictionary);
            WebUtils web = new WebUtils();
            web.HttpRequest(requestUri, HttpMethod.GET, WebUtils.BASIC_TIMEOUT); // Nothing returned
        }

        /* Updates password */
        public void UpdatePassword(Guid apiToken, string accountPassword)
        {
            this.AssertArgument(accountPassword, "New Password");

            Dictionary<string, string> queryDictionary = this.InitAccessReqDict(apiToken);
            queryDictionary.Add("account_password", accountPassword);

            Uri requestUri = Utils.BuildUri(BASE_URI, UPDATE_PASSWORD_PATH, queryDictionary);
            WebUtils web = new WebUtils();

            // TODO: returns a MISSING_PARAMETERS error, although everything is supplied

            web.HttpRequest(requestUri, HttpMethod.POST, WebUtils.BASIC_TIMEOUT); // Nothing returned
        }

        /* Returns a new Secure API key */
        public Guid GenerateAPIKey(Guid apiToken, string username, bool forceNew = false)
        {
            this.AssertArgument(username, "Username/Account ID");

            Dictionary<string, string> queryDictionary = this.InitAccessReqDict(apiToken);
            queryDictionary.Add("account_id", username);
            queryDictionary.Add("force_new", forceNew.ToString());

            Uri requestUri = Utils.BuildUri(BASE_URI, GENERATE_API_KEY_PATH, queryDictionary);
            WebUtils web = new WebUtils();

            string serverResponse = web.HttpRequest(requestUri, HttpMethod.GET, WebUtils.BASIC_TIMEOUT);
            Dictionary<string, string> response = Utils.DeserializeDictionary(serverResponse);

            return new Guid(response["ApiKey"]);
        }

        /* Deactivates the supplied Secure API key */
        public void RemoveAPIKey(Guid apiToken, Guid apiSecurekey)
        {
            Dictionary<string, string> queryDictionary = this.InitAccessReqDict(apiToken);
            queryDictionary.Add("api_securekey", apiSecurekey.ToString("N"));

            Uri requestUri = Utils.BuildUri(BASE_URI, REMOVE_API_KEY_PATH, queryDictionary);
            WebUtils web = new WebUtils();
            web.HttpRequest(requestUri, HttpMethod.GET, WebUtils.BASIC_TIMEOUT); // Nothing returned
        }


        /// JOB CONTROL ///
        
        /* Creates a new job. Returns an array of Guids where 'JobId' is the 0th element and 'TaskId' is the 1st element */
        public Guid[] CreateJob(Guid apiToken, string jobName = null, string sourceLanguage = "en")
        {
            Dictionary<string, string> queryDictionary = this.InitAccessReqDict(apiToken);
            if (jobName != null) { queryDictionary.Add("job_name", jobName); }
            queryDictionary.Add("source_language", sourceLanguage);

            Uri requestUri = Utils.BuildUri(BASE_URI, CREATE_JOB_PATH, queryDictionary);
            WebUtils web = new WebUtils();

            string serverResponse = web.HttpRequest(requestUri, HttpMethod.GET, WebUtils.BASIC_TIMEOUT);
            Dictionary<string, string> response = Utils.DeserializeDictionary(serverResponse);

            Guid jobId = new Guid(response["JobId"].ToString());
            Guid taskId = new Guid(response["TaskId"].ToString());

            return new Guid[] { jobId, taskId };
        }

        /* Authorizes a job with jobId */
        public void AuthorizeJob(Guid apiToken, Guid jobId)
        {
            Dictionary<string, string> queryDictionary = InitJobReqDict(apiToken, jobId);

            Uri requestUri = Utils.BuildUri(BASE_URI, AUTHORIZE_JOB_PATH, queryDictionary);
            WebUtils web = new WebUtils();

            web.HttpRequest(requestUri, HttpMethod.GET, WebUtils.BASIC_TIMEOUT); // Nothing returned
        }

        /* Deletes a job with jobId */
        public Guid DeleteJob(Guid apiToken, Guid jobId)
        {
            Dictionary<string, string> queryDictionary = InitJobReqDict(apiToken, jobId);

            Uri requestUri = Utils.BuildUri(BASE_URI, DELETE_JOB_PATH, queryDictionary);
            WebUtils web = new WebUtils();

            string serverResponse = web.HttpRequest(requestUri, HttpMethod.GET, WebUtils.BASIC_TIMEOUT);
            
            // TODO: server returns Guid as a raw string, not inside of a json object

            Dictionary<string, string> response = Utils.DeserializeDictionary(serverResponse);

            return new Guid(response["TaskId"]);
        }

        public string GetJobInfo(Guid apiToken, Guid jobId)
        {
            Dictionary<string, string> queryDictionary = InitJobReqDict(apiToken, jobId);

            Uri requestUri = Utils.BuildUri(BASE_URI, GET_JOB_INFO_PATH, queryDictionary);
            WebUtils web = new WebUtils();

            string serverResponse = web.HttpRequest(requestUri, HttpMethod.GET, WebUtils.BASIC_TIMEOUT);
            Dictionary<string, string> response = Utils.DeserializeDictionary(serverResponse);

            // TODO

            return null;
        }

        public string GetJobList(Guid apiToken)
        {            
            Dictionary<string, string> queryDictionary = new Dictionary<string, string>();
            queryDictionary.Add("v", VERSION.ToString());
            queryDictionary.Add("api_token", apiToken.ToString());

            Uri requestUri = Utils.BuildUri(BASE_URI, GET_JOB_LIST_PATH, queryDictionary);
            WebUtils web = new WebUtils();

            string serverResponse = web.HttpRequest(requestUri, HttpMethod.GET, WebUtils.BASIC_TIMEOUT);
            Dictionary<string, string> response = Utils.DeserializeDictionary(serverResponse);

            // TODO

            return null;
        }

        public Guid AddMediaToJob(Guid apiToken, Guid jobId, Uri mediaUrl)
        {
            if (mediaUrl == null || mediaUrl.ToString().Equals("")) { throw new ArgumentException("Invalid Media URL"); }

            Dictionary<string, string> queryDictionary = InitJobReqDict(apiToken, jobId);
            queryDictionary.Add("media_url", Utils.EncodeUrl(mediaUrl));

            Uri requestUri = Utils.BuildUri(BASE_URI, ADD_MEDIA_TO_JOB_PATH, queryDictionary);
            WebUtils web = new WebUtils();

            string serverResponse = web.HttpRequest(requestUri, HttpMethod.GET, WebUtils.BASIC_TIMEOUT);
            Dictionary<string, string> response = Utils.DeserializeDictionary(serverResponse);

            return new Guid(response["TaskId"]);
        }

        public Guid AddMediaToJob(Guid apiToken, Guid jobId, Stream fileStream)
        {
            if (fileStream == null) { throw new ArgumentException("Invalid File"); }

            Dictionary<string, string> queryDictionary = InitJobReqDict(apiToken, jobId);
            Uri requestUri = Utils.BuildUri(BASE_URI, ADD_MEDIA_TO_JOB_PATH, queryDictionary);
            WebUtils web = new WebUtils();

            // TODO: Content-Type: video/mp4 ???
            string serverResponse = web.UploadMedia(requestUri, fileStream, "video/mp4");
            Dictionary<string, string> response = Utils.DeserializeDictionary(serverResponse);
            
            return new Guid(response["TaskId"]);
        }

        public Guid AddEmbeddedMediaToJob(Guid apiToken, Guid jobId, Uri mediaUrl)
        {
            if (mediaUrl == null || mediaUrl.ToString().Equals("")) { throw new ArgumentException("Invalid Media URL"); }

            Dictionary<string, string> queryDictionary = InitJobReqDict(apiToken, jobId);
            queryDictionary.Add("media_url", Utils.EncodeUrl(mediaUrl));

            Uri requestUri = Utils.BuildUri(BASE_URI, ADD_EMBEDDED_MEDIA_TO_JOB_PATH, queryDictionary);
            WebUtils web = new WebUtils();

            string serverResponse = web.HttpRequest(requestUri, HttpMethod.GET, WebUtils.BASIC_TIMEOUT);
            Dictionary<string, string> response = Utils.DeserializeDictionary(serverResponse);

            return new Guid(response["TaskId"]);
        }

        public Uri GetMedia(Guid apiToken, Guid jobId)
        {
            Dictionary<string, string> queryDictionary = InitJobReqDict(apiToken, jobId);

            Uri requestUri = Utils.BuildUri(BASE_URI, GET_MEDIA_PATH, queryDictionary);
            WebUtils web = new WebUtils();

            string serverResponse = web.HttpRequest(requestUri, HttpMethod.GET, WebUtils.BASIC_TIMEOUT);
            Dictionary<string, string> response = Utils.DeserializeDictionary(serverResponse);

            // TODO: Url that's being returned does not match the Url used in AddMediaToJob. Returns "api.sandbox.cielo24.com"

            return new Uri(Utils.UnescapeString(response["MediaUrl"]));
        }

        public string GetTranscript(Guid apiToken, Guid jobId, CaptionOptions captionOptions = null)
        {
            Dictionary<string, string> queryDictionary = InitJobReqDict(apiToken, jobId);
            if (captionOptions != null) { queryDictionary.Add("caption_options", captionOptions.ToString()); }

            Uri requestUri = Utils.BuildUri(BASE_URI, GET_TRANSCRIPTION_PATH, queryDictionary);
            WebUtils web = new WebUtils();

            return web.HttpRequest(requestUri, HttpMethod.GET, WebUtils.DOWNLOAD_TIMEOUT); // Transcript text
        }

        public string GetCaption(Guid apiToken, Guid jobId, CaptionFormat captionFormat=CaptionFormat.SRT, CaptionOptions captionOptions = null)
        {
            Dictionary<string, string> queryDictionary = InitJobReqDict(apiToken, jobId);
            queryDictionary.Add("caption_format", captionFormat.ToString());
            if (captionOptions != null) { queryDictionary.Add("caption_options", captionOptions.ToString()); }

            Uri requestUri = Utils.BuildUri(BASE_URI, GET_CAPTION_PATH, queryDictionary);
            WebUtils web = new WebUtils();

            return web.HttpRequest(requestUri, HttpMethod.GET, WebUtils.DOWNLOAD_TIMEOUT); // Caption text
        }

        public string GetElementList(Guid apiToken, Guid jobId)
        {
            Dictionary<string, string> queryDictionary = InitJobReqDict(apiToken, jobId);

            Uri requestUri = Utils.BuildUri(BASE_URI, GET_ELEMENT_LIST_PATH, queryDictionary);
            WebUtils web = new WebUtils();

            string serverResponse = web.HttpRequest(requestUri, HttpMethod.GET, WebUtils.BASIC_TIMEOUT);
            Dictionary<string, string> response = Utils.DeserializeDictionary(serverResponse);
            
            // TODO

            return null;
        }

        public string GetListOfElementLists(Guid apiToken, Guid jobId)
        {
            Dictionary<string, string> queryDictionary = InitJobReqDict(apiToken, jobId);

            Uri requestUri = Utils.BuildUri(BASE_URI, GET_LIST_OF_ELEMENT_LISTS_PATH, queryDictionary);
            WebUtils web = new WebUtils();

            string serverResponse = web.HttpRequest(requestUri, HttpMethod.GET, WebUtils.BASIC_TIMEOUT);
            Dictionary<string, string> response = Utils.DeserializeDictionary(serverResponse);

            // TODO

            return null;
        }


        /// PRIVATE HELPER METHODS ///
        
        /* Returns a dictionary with version, api_token and job_id key-value pairs (parameters used in almost every job-control action). */
        private Dictionary<string, string> InitJobReqDict(Guid apiToken, Guid jobId)
        {
            // Guids are non-nullable. No need to assert.
            Dictionary<string, string> queryDictionary = this.InitAccessReqDict(apiToken);
            queryDictionary.Add("job_id", jobId.ToString("N"));
            return queryDictionary;
        }

        /* Returns a dictionary with version and api_token key-value pairs (parameters used in almst every access-control action). */
        private Dictionary<string, string> InitAccessReqDict(Guid apiToken)
        {
            Dictionary<string, string> queryDictionary = new Dictionary<string, string>();
            queryDictionary.Add("v", VERSION.ToString());
            queryDictionary.Add("api_token", apiToken.ToString("N"));
            return queryDictionary;
        }

        /* If arg is invalid (null or empty), throws an ArgumentException */
        private void AssertArgument(string arg, string arg_name)
        {
            if (arg == null || arg.Equals("")) { throw new ArgumentException("Invalid " + arg_name); }
        }
    }
}

File: cielo24_PHP/src/cielo24/Actions.php
<?php

class Actions {

    public $BASE_URL = "https://api.cielo24.com";
    const API_VERSION = 1;

    const LOGIN_PATH = "/api/account/login";
    const LOGOUT_PATH = "/api/account/logout";
    const UPDATE_PASSWORD_PATH = "/api/account/update_password";
    const GENERATE_API_KEY_PATH = "/api/account/generate_api_key";
    const REMOVE_API_KEY_PATH = "/api/account/remove_api_key";
    const CREATE_JOB_PATH = "/api/job/new";
    const AUTHORIZE_JOB_PATH = "/api/job/authorize";
    const DELETE_JOB_PATH = "/api/job/del";
    const GET_JOB_INFO_PATH = "/api/job/info";
    const GET_JOB_LIST_PATH = "/api/job/list";
    const ADD_MEDIA_TO_JOB_PATH = "/api/job/add_media";
    const ADD_EMBEDDED_MEDIA_TO_JOB_PATH = "/api/job/add_media_url";
    const GET_MEDIA_PATH = "/api/job/media";
    const PERFORM_TRANSCRIPTION = "/api/job/perform_transcription";
    const GET_TRANSCRIPTION_PATH = "/api/job/get_transcript";
    const GET_CAPTION_PATH = "/api/job/get_caption";
    const GET_ELEMENT_LIST_PATH = "/api/job/get_elementlist";
    const GET_LIST_OF_ELEMENT_LISTS_PATH = "/api/job/list_elementlists";

    public function __construct() { }

    public function __construct1($base_url) {
        $this->BASE_URL = $base_url;
    }

    public function __call($name, $arguments) {
        // TODO
    }

    /// ACCESS CONTROL ///

    /* Performs a Login action. If useHeaders is true, puts username and password into HTTP headers */
    public function login($username, $password = null, $api_securekey = null, $use_headers = false) {
        $this->_assertArgument($username, "Username");

        # Password or API Secure Key must be supplied but not both
        if ($password == null and $api_securekey == null) {
            throw new InvalidArgumentException("Password or API Secure Key must be supplied for login.");
        }

        $query_dict = $this->_initVersionDict();
        $headers = array();

        if ($use_headers) {
            $headers["x-auth-user"] = $username;
            if ($password != null) {
                $headers["x-auth-key"] = $password;
            }
            if ($api_securekey != null) {
                $headers["x-auth-securekey"] = $api_securekey;
            }
        } else {
            $query_dict["username"] = $username;
            if ($password != null) {
                $query_dict["password"] = $password;
            }
            if ($api_securekey != null) {
                $query_dict["securekey"] = $api_securekey;
            }
        }

        $response = WebUtils::getJson($this->BASE_URL, Actions::LOGIN_PATH, HTTP_METH_GET, WebUtils::BASIC_TIMEOUT, $query_dict, $headers);
        return $response["ApiToken"];
    }

    /* Performs a Logout action */
    public function logout($api_token) {
        $query_dict = $this->_initAccessReqDict($api_token);
        // Nothing returned
        WebUtils::httpRequest($this->BASE_URL, Actions::LOGOUT_PATH, HTTP_METH_GET, WebUtils::BASIC_TIMEOUT, $query_dict);
    }

    /* Updates password */
    public function updatePassword($api_token, $new_password) {
        $this->_assertArgument($new_password, "New Password");

        $query_dict = $this->_initAccessReqDict($api_token);
        $query_dict["new_password"] = $new_password;
        // Nothing returned
        WebUtils::httpRequest($this->BASE_URL, Actions::UPDATE_PASSWORD_PATH, HTTP_METH_POST, WebUtils::BASIC_TIMEOUT, null, null, http_build_query($query_dict));
    }

    /* Returns a new Secure API key */
    public function generateAPIKey($api_token, $username, $force_new = false) {
        $this->_assertArgument($username, "Username");

        $query_dict = $this->_initAccessReqDict($api_token);
        $query_dict["account_id"] = $username;
        $query_dict["force_new"] = ($force_new) ? 'true' : 'false';

        $response = WebUtils::getJson($this->BASE_URL, Actions::GENERATE_API_KEY_PATH, HTTP_METH_GET, WebUtils::BASIC_TIMEOUT, $query_dict);
        return $response["ApiKey"];
    }

    /* Deactivates the supplied Secure API key */
    public function removeAPIKey($api_token, $api_securekey) {
        $query_dict = $this->_initAccessReqDict($api_token);
        $query_dict["api_securekey"] = $api_securekey;
        // Nothing returned
        WebUtils::httpRequest($this->BASE_URL, Actions::REMOVE_API_KEY_PATH, HTTP_METH_GET, WebUtils::BASIC_TIMEOUT, $query_dict);
    }

    /// JOB CONTROL ///

    /* Creates a new job. Returns an array of Guids where 'JobId' is the 0th element and 'TaskId' is the 1st element */
    public function createJob($api_token, $job_name = null, $language = "en") {
        $query_dict = $this->_initAccessReqDict($api_token);
        if($job_name) {
            $query_dict["job_name"] = $job_name;
        }
        $query_dict["language"] = ($language == null) ? "en" : $language;

        $response = WebUtils::getJson($this->BASE_URL, Actions::CREATE_JOB_PATH, HTTP_METH_GET, WebUtils::BASIC_TIMEOUT, $query_dict);
        // Return a hash with JobId and TaskId
        return $response;
    }

    /* Authorizes a job with job_id */
    public function authorizeJob($api_token, $job_id) {
        $query_dict = $this->_initJobReqDict($api_token, $job_id);
        // Nothing returned\
        WebUtils::httpRequest($this->BASE_URL, Actions::AUTHORIZE_JOB_PATH, HTTP_METH_GET, WebUtils::BASIC_TIMEOUT, $query_dict);
    }

    /* Deletes a job with job_id */
    public function deleteJob($api_token, $job_id) {
        $query_dict = $this->_initJobReqDict($api_token, $job_id);
        $response = WebUtils::getJson($this->BASE_URL, Actions::DELETE_JOB_PATH, HTTP_METH_GET, WebUtils::BASIC_TIMEOUT, $query_dict);
        return $response["TaskId"];
    }

    /* Gets information about a job with job_id */
    public function getJobInfo($api_token, $job_id) {
        $query_dict = $this->_initJobReqDict($api_token, $job_id);
        $response = WebUtils::getJson($this->BASE_URL, Actions::GET_JOB_INFO_PATH, HTTP_METH_GET, WebUtils::BASIC_TIMEOUT, $query_dict);
        return $response;
    }

    /* Gets a list of jobs */
    public function getJobList($api_token) {
        $query_dict = $this->_initAccessReqDict($api_token);
        $response = WebUtils::getJson($this->BASE_URL, Actions::GET_JOB_LIST_PATH, HTTP_METH_GET, WebUtils::BASIC_TIMEOUT, $query_dict);
        return $response;
    }

    /* Uploads a file from fileStream to job with job_id */
    public function addMediaToJobFile($api_token, $job_id, $file_stream) {


    }

    /* Provides job with job_id a url to media */
    public function addMediaToJobUrl($api_token, $job_id, $media_url) {

    }

    /* Provides job with job_id a url to media */
    public function addMediaToJobEmbedded($api_token, $job_id, $media_url) {

    }

    /* Helper method for AddMediaToJob and AddEmbeddedMediaToJob methods */
    private function _sendMediaUrl($api_token, $job_id, $media_url, $path) {
        $this->_assertArgument($media_url, "Media URL");
        $query_dict = $this->_initJobReqDict($api_token, $job_id);
        $query_dict['media_url'] = $media_url;

        $response = WebUtils::getJson($this->BASE_URL, $path, HTTP_METH_GET, WebUtils::BASIC_TIMEOUT, $query_dict);
        return $response["TaskId"];
    }

    /* Returns a Uri to the media from job with job_id */
    public function getMedia($api_token, $job_id) {
        $query_dict = $this->_initJobReqDict($api_token, $job_id);
        $response = WebUtils::getJson($this->BASE_URL, Actions::GET_MEDIA_PATH, HTTP_METH_GET, WebUtils::BASIC_TIMEOUT, $query_dict);
        return $response["MediaUrl"];
    }

    /* Makes a PerformTranscription call */
    public function perform_transcription($api_token,
                                         $job_id,
                                         $fidelity,
                                         $priority,
                                         $callback_uri = null,
                                         $turnaround_hours = null,
                                         $target_language = null,
                                         $options = null) {

    }

    /* Returns a transcript from a job with job_id */
    public function getTranscript($api_token, $job_id, $transcript_options = null) {
        $query_dict = $this->_initJobReqDict($api_token, $job_id);
        if ($transcript_options != null) {
            $query_dict += $transcript_options;
        }
        # Return raw transcript text
        return WebUtils::httpRequest($this->BASE_URL, Actions::GET_TRANSCRIPTION_PATH, HTTP_METH_GET, WebUtils::BASIC_TIMEOUT, $query_dict);
    }

    /* Returns a caption from a job with job_id OR if buildUri is true, returns a string representation of the uri */
    public function getCaption($api_token, $job_id, $caption_format, $caption_options = null) {

    }

    /* Returns an element list */
    public function getElementList($api_token, $job_id, $element_list_version = null) {

    }

    /* Returns a list of elements lists */
    public function getListOfElementLists($api_token, $job_id) {

    }

    /// PRIVATE HELPER METHODS ///

    private function get_job_response($api_token, $job_id, $path) {
        return null;
    }

    /* Returns a dictionary with version, api_token and job_id key-value pairs (parameters used in almost every job-control action). */
    private function _initJobReqDict($api_token, $job_id) {
        $this->_assertArgument($api_token, "Job ID");
        $dict = $this->_initAccessReqDict($api_token);
        $dict["job_id"] = $job_id;
        return $dict;
    }

    /* Returns a dictionary with version and api_token key-value pairs (parameters used in almost every access-control action). */
    private function _initAccessReqDict($api_token) {
        $this->_assertArgument($api_token, "API Token");
        $dict = $this->_initVersionDict();
        $dict["api_token"] = $api_token;
        return $dict;
    }

    /* Returns a dictionary with version key-value pair (parameter used in every action). */
    private function _initVersionDict() {
        return array("v" => Actions::API_VERSION);
    }

    /* If arg is invalid (null or empty), throws an InvalidArgumentException */
    private function _assertArgument($arg, $arg_name) {
        if ($arg == null) { throw new InvalidArgumentException("Invalid " . $arg_name); }
        elseif (gettype($arg) == "string" and strlen($arg) == 0) { throw new InvalidArgumentException("Invalid " . $arg_name); }
    }
}